File: packages/agents/src/claude-agent.ts
/**
 * `kriya-agents/claude-agent` — govern a Claude Agent SDK (TypeScript) custom tool with kriya.
 *
 * The Claude Agent SDK defines an in-process tool with
 * `tool(name, description, inputSchema, handler)`, where `handler(args, extra) => { content, isError? }`
 * (verified 2026-07-21 against the Claude Agent SDK TS reference). Governing it is governing that
 * `handler` — this adapter is a thin shim over the framework-agnostic {@link govern} with no
 * dependency on `@anthropic-ai/claude-agent-sdk`: you wrap your handler, then pass it to the SDK's
 * `tool()` yourself, so name/description/schema (what the model sees) are unchanged.
 *
 * ```ts
 * import { tool } from "@anthropic-ai/claude-agent-sdk";
 * import { z } from "zod";
 * import { GovernClient } from "kriya-agents";
 * import { governClaudeHandler } from "kriya-agents/claude-agent";
 *
 * const client = new GovernClient({ actor: "claude-agent", auditLog: "…/.kriya/audit/claude.jsonl" });
 * const search = tool(
 *   "search", "Search the web", { query: z.string() },
 *   governClaudeHandler(client, "search", async ({ query }) => ({
 *     content: [{ type: "text", text: `Results for: ${query}` }],
 *   })),
 * );
 * ```
 *
 * Honest ceiling (state it in your app too): in-process governance is cooperative — a hostile agent
 * process can skip it (that is what launch-under containment / B14 is for) — and this lane governs the
 * action tier + signs the receipt; it does not see the tool's own outbound network calls.
 */

import { type GovernClient, type GovernHooks, govern } from "./govern-client.js";

/** The `CallToolResult` a Claude Agent SDK tool handler returns. */
export interface ClaudeToolResult {
  content: unknown[];
  isError?: boolean;
  [key: string]: unknown;
}

/** A Claude Agent SDK tool handler: `(args, extra) => CallToolResult`. */
export type ClaudeToolHandler<A> = (args: A, extra: unknown) => ClaudeToolResult | Promise<ClaudeToolResult>;

/**
 * Wrap a Claude Agent SDK tool handler so every invocation is policy / approval / budget gated and
 * (when it runs) signs a receipt. `actionId` is the id policy rules match on — use the tool's name.
 * A denied / approval-refused / over-budget call returns an `isError` `CallToolResult` (so the model
 * sees the tool errored and adapts) — not a thrown exception, which is the idiomatic MCP tool shape.
 */
export function governClaudeHandler<A extends Record<string, unknown>>(
  client: GovernClient,
  actionId: string,
  handler: ClaudeToolHandler<A>,
  hooks?: GovernHooks,
): ClaudeToolHandler<A> {
  // Reuse the framework-agnostic core for the check/record + receipt semantics, then map its
  // GovernDenied throw onto the SDK's `{ isError: true }` result shape.
  const run = govern<A, ClaudeToolResult>(
    client,
    actionId,
    // The wrapped "tool" is the handler; its `isError` result counts as a failed call for the receipt.
    async (args: A) => {
      const result = await handler(args, undefined);
      if (result.isError) {
        // Signal failure to `govern` (so the receipt records success:false) without losing the
        // handler's own error content — re-thrown and re-wrapped below.
        throw new ClaudeToolError(result);
      }
      return result;
    },
    hooks,
  );
  return async (args: A, _extra: unknown): Promise<ClaudeToolResult> => {
    try {
      return await run(args);
    } catch (err) {
      if (err instanceof ClaudeToolError) return err.result; // the handler's own isError result
      // A GovernDenied (or any other pre-execution error) → an isError result the model can adapt to.
      const text = err instanceof Error ? err.message : String(err);
      return { content: [{ type: "text", text }], isError: true };
    }
  };
}

/** Internal: carries a handler's own `isError` result through `govern`'s failure path. */
class ClaudeToolError extends Error {
  constructor(readonly result: ClaudeToolResult) {
    super("claude tool returned isError");
    this.name = "ClaudeToolError";
  }
}

File: bindings/python/src/kriya/agents/claude_agent.py
"""kriya.agents.claude_agent — govern a Claude Agent SDK (Python) custom tool with kriya.

The Claude Agent SDK (Python) defines an in-process tool with the `@tool(name, description,
input_schema)` decorator over an async `handler(args) -> {"content": [...], "isError"?}` (verified
2026-07-22 against `claude-agent-sdk` 0.2.125). Governing the tool is governing that handler: this
adapter wraps the built `SdkMcpTool.handler` so every call is policy -> approval -> budget gated and
(when it runs) signs a receipt, with name/description/input_schema untouched. No dependency on
`claude_agent_sdk` -- you build the tool, then hand it to `govern_sdk_tool`.

```python
from claude_agent_sdk import tool, create_sdk_mcp_server
from kriya.agents import GovernClient
from kriya.agents.claude_agent import govern_sdk_tool

client = GovernClient(actor="claude-agent", audit_log="~/.kriya/audit/claude.jsonl")

@tool("search", "Search the web", {"query": str})
async def search(args):
    return {"content": [{"type": "text", "text": f"Results for: {args['query']}"}]}

govern_sdk_tool(client, search)   # same tool object, now governed
server = create_sdk_mcp_server(name="my-tools", tools=[search])
```

A denied / approval-refused / over-budget call returns an `isError` result (idiomatic MCP tool shape
-- the model sees the tool errored and adapts), not a raised exception. Honest ceiling: in-process
governance is cooperative (a hostile process can skip it -- that is what launch-under containment /
B14 is for); it governs the action tier + signs, and does not see the tool's own egress.
"""

from __future__ import annotations

from typing import Any, Dict, Optional

from . import GovernClient

__all__ = ["govern_sdk_tool"]


def _denied_result(action_id: str, decision: str, reason: Optional[str]) -> Dict[str, Any]:
    detail = f" — {reason}" if reason else ""
    return {
        "content": [{"type": "text", "text": f'kriya denied "{action_id}": {decision}{detail}'}],
        "isError": True,
    }


def govern_sdk_tool(client: GovernClient, tool: Any, action_id: Optional[str] = None) -> Any:
    """Govern a built Claude Agent SDK `SdkMcpTool` in place (its `handler` is wrapped). Returns the
    same `tool`. `action_id` defaults to `tool.name` -- the id your policy rules match on.
    """
    resolved_id = action_id or getattr(tool, "name", None)
    if not resolved_id:
        raise ValueError("govern_sdk_tool: tool has no name and no action_id was given")
    original = tool.handler

    async def governed(args: Any) -> Dict[str, Any]:
        params = dict(args) if isinstance(args, dict) else {"input": args}
        gate = client.check(resolved_id, params)
        if gate.decision != "allow":
            # Action-tier deny signs no receipt (parity with the governor); surface it as isError.
            return _denied_result(resolved_id, gate.decision, gate.reason)
        success = False
        try:
            result = await original(args)
            success = not (isinstance(result, dict) and result.get("isError"))
            return result
        finally:
            client.record(resolved_id, params, success)

    tool.handler = governed
    return tool
